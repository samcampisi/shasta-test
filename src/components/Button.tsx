import React from 'react';
import {
  StyleProp,
  ViewStyle,
  TouchableOpacity,
  Text,
  ImagePropertiesSourceOptions,
  Image,
  ImageStyle,
} from 'react-native';
import styles from '../styles/Button.style';

export interface ButtonProps {
  title: string;
  disabled?: boolean;
  onPress?: () => void;
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<ViewStyle>;
  imageStyle?: StyleProp<ImageStyle>;
  icon?: ImagePropertiesSourceOptions;
  testID?: string;
}

const Button = (props: ButtonProps) => {
  const { title, disabled, onPress, style, textStyle, icon, imageStyle, testID } = props;

  const onButtonPress = () => {
    !disabled && onPress && onPress();
  };

  return (
    <TouchableOpacity
      testID={testID}
      style={[styles.container, disabled && styles.disabled, style]}
      disabled={disabled}
      onPress={onButtonPress}>
      {icon && (
        <Image
          source={icon}
          style={[styles.icon, imageStyle]}
          resizeMode="contain"
          testID="button-icon"
        />
      )}
      <Text style={[styles.text, textStyle]}>{title}</Text>
    </TouchableOpacity>
  );
};

export default Button;
