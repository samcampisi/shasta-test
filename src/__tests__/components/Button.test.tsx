import Button from '../../components/Button';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { render, fireEvent } from 'react-native-testing-library';

Enzyme.configure({ adapter: new Adapter() });

const createTestProps = (props?: object) => ({
  ...props,
});

describe('Button', () => {
  const props: any = createTestProps({
    onPress: jest.fn(),
    title: 'Test title',
  });

  describe('Basic button', () => {
    const { getByText, toJSON } = render(<Button {...props} />);

    it('should render the title', () => {
      expect(getByText('Test title')).toBeDefined();
    });

    it('should match snapshot', () => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('Button with custom style', () => {
    const { toJSON } = render(<Button {...props} style={{ backgroundColor: 'yellow' }} />);

    it('should match snapshot', () => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('Button with custom icon', () => {
    const { getByTestId, toJSON } = render(
      <Button
        {...props}
        icon={{ source: 'https://d2ph5fj80uercy.cloudfront.net/06/cat1996.jpg' }}
      />,
    );

    it('should render the title', () => {
      expect(getByTestId('button-icon')).toBeDefined();
    });

    it('should match snapshot', () => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('Disabled', () => {
    const { getByText, toJSON } = render(<Button {...props} disabled />);

    fireEvent.press(getByText('Test title'));

    it('should match snapshot', () => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
