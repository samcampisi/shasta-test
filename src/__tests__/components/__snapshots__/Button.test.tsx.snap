// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Basic button should match snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#ff66a4",
      "borderRadius": 15,
      "flexDirection": "row",
      "margin": 15,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Test title
  </Text>
</View>
`;

exports[`Button Button with custom icon should match snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#ff66a4",
      "borderRadius": 15,
      "flexDirection": "row",
      "margin": 15,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "source": "https://d2ph5fj80uercy.cloudfront.net/06/cat1996.jpg",
      }
    }
    style={
      Array [
        Object {
          "height": 25,
          "marginRight": 10,
          "tintColor": "#ffffff",
          "width": 30,
        },
        undefined,
      ]
    }
    testID="button-icon"
  />
  <Text
    style={
      Array [
        Object {
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Test title
  </Text>
</View>
`;

exports[`Button Button with custom style should match snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "yellow",
      "borderRadius": 15,
      "flexDirection": "row",
      "margin": 15,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Test title
  </Text>
</View>
`;

exports[`Button Disabled should match snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#ff66a4",
      "borderRadius": 15,
      "flexDirection": "row",
      "margin": 15,
      "opacity": 0.5,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "600",
        },
        undefined,
      ]
    }
  >
    Test title
  </Text>
</View>
`;
